src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,331 @@
+      Vector<BaseFloat> early_reverb(signal);
+      BlockConvolveSignals(early_rir, &early_reverb);
+
+      // compute the energy
+      return VecVec(early_reverb, early_reverb) / early_reverb.Dim();
+}
+}
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+
+    const char *usage =
+        "Corrupts the wave files supplied via input pipe with the specified\n"
+        "room-impulse response (rir_matrix) and additive noise distortions\n"
+        "(specified by corresponding files).\n";
+
+    ParseOptions po(usage);
+    std::string noise_file;
+    BaseFloat snr_db = 20;
+    bool test_mode = false;
+    bool multi_channel_output = false;
+    int input_wave_channel = 0;
+    int rir_channel = 0;
+    int noise_channel = 0;
+
+    po.Register("multi-channel-output", &multi_channel_output,
+                "Output with multiple channels");
+    po.Register("input-wave-channel", &input_wave_channel,
+                "specify the channel of the input");
+    po.Register("rir-channel", &rir_channel,
+                "specify the channel of the RIR");
+    po.Register("noise-channel", &noise_channel,
+                "specify the channel of the noise");
+    po.Register("test-mode", &test_mode,
+                "Add a small unit test to confirm that the "
+                "block convolution program generates the "
+                "same results as the full signal convolution");
+    po.Register("noise-file", &noise_file,
+                "file with additive noise");
+    po.Register("snr-db", &snr_db,
+                "desired SNR(dB) of the output");
+
+    po.Read(argc, argv);
+    if (po.NumArgs() != 3) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    std::string input_wave_file = po.GetArg(1);
+    std::string rir_file = po.GetArg(2);
+    std::string output_wave_file = po.GetArg(3);
+
+    WaveData input_wave;
+    {
+      Input ki(input_wave_file);
+      input_wave.Read(ki.Stream());
+    }
+
+    const Matrix<BaseFloat> &input_matrix = input_wave.Data();
+    BaseFloat samp_freq_input = input_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_input = input_matrix.NumCols(),  // number of samples in recording
+          num_chan_input = input_matrix.NumRows();  // number of channels in recording
+
+    KALDI_LOG << "samp_freq of source is " << samp_freq_input
+              << " num_samp " << num_samp_input
+              << " num_chan " << num_chan_input;
+
+    KALDI_ASSERT(input_wave_channel < num_chan_input);
+
+    WaveData rir_wave;
+    {
+      Input ki(rir_file);
+      rir_wave.Read(ki.Stream());
+    }
+    const Matrix<BaseFloat> &rir_matrix = rir_wave.Data();
+    BaseFloat samp_freq_rir = rir_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_rir = rir_matrix.NumCols(),  // number of samples in recording
+          num_chan_rir = rir_matrix.NumRows();  // number of channels in recording
+    KALDI_LOG << "samp_freq of rir is " << samp_freq_rir
```

Comment:
You should eliminate these logging statements or at least move them to the debug log. Same for noise and input wave files. 

You should reduce the size of the main function, it is 170 lines. You should keep it around 40. This makes it easier for reading. You could move everything other than file reading and corresponding asserts to new functions. e.g. create a new function for computing each channel output in the while loop @ line 273.
Move the code under testmode to a function like UnitTestBlockConvolution.